oracle/oracles/pd.go

```diff
@@ -241,28 +345,172 @@ func (o *pdOracle) getLastTSWithArrivalTS(txnScope string) (*lastTSO, bool) {
+		// the user's configured value directly.
+		if configuredInterval <= minAllowedAdaptiveUpdateTSInterval {
+			return adaptiveUpdateTSIntervalStateUnadjustable, configuredInterval
+		}
+		return none, 0
+	}
+
+	checkNormal := func() (adaptiveUpdateTSIntervalState, time.Duration) {
+		// If the current actual update interval is synced with the configured value, and it's not unadjustable state,
+		// then it's the normal state.
+		if configuredInterval > minAllowedAdaptiveUpdateTSInterval && currentAdaptiveUpdateInterval == configuredInterval {
+			return adaptiveUpdateTSIntervalStateNormal, currentAdaptiveUpdateInterval
+		}
+		return none, 0
+	}
+
+	checkAdapting := func() (adaptiveUpdateTSIntervalState, time.Duration) {
+		if requiredStaleness != 0 && requiredStaleness < currentAdaptiveUpdateInterval && currentAdaptiveUpdateInterval > minAllowedAdaptiveUpdateTSInterval {
+			// If we are calculating the interval because of a request that requires a shorter staleness, we shrink the
+			// update interval immediately to adapt to it.
+			// We shrink the update interval to a value slightly lower than the requested staleness to avoid potential
+			// frequent shrinking operations. But there's a lower bound to prevent loading ts too frequently.
+			newInterval := max(requiredStaleness-adaptiveUpdateTSIntervalShrinkingPreserve, minAllowedAdaptiveUpdateTSInterval)
+			return adaptiveUpdateTSIntervalStateAdapting, newInterval
+		}
+
+		if currentAdaptiveUpdateInterval != configuredInterval && now.Sub(lastReachDropThresholdTime) < adaptiveUpdateTSIntervalDelayBeforeRecovering {
+			// There is a recent request that requires a short staleness. Keep the current adaptive interval.
+			// If it's not adapting state, it's possible that it's previously in recovering state, and it stops recovering
+			// as there is a new read operation requesting a short staleness.
+			return adaptiveUpdateTSIntervalStateAdapting, currentAdaptiveUpdateInterval
+		}
+
+		return none, 0
+	}
+
+	checkRecovering := func() (adaptiveUpdateTSIntervalState, time.Duration) {
+		if currentAdaptiveUpdateInterval == configuredInterval || now.Sub(lastReachDropThresholdTime) < adaptiveUpdateTSIntervalDelayBeforeRecovering {
+			return none, 0
+		}
+
+		timeSinceLastTick := now.Sub(o.adaptiveUpdateIntervalState.lastTick)
+		newInterval := currentAdaptiveUpdateInterval + time.Duration(timeSinceLastTick.Seconds()*float64(adaptiveUpdateTSIntervalRecoverPerSecond))
+		if newInterval > configuredInterval {
+			newInterval = configuredInterval
+		}
+
+		return adaptiveUpdateTSIntervalStateRecovering, newInterval
+	}
+
+	// Check the specified states in order, until the state becomes determined.
+	// If it's still undetermined after all checks, keep the previous state.
+	nextState := func(checkFuncs ...func() (adaptiveUpdateTSIntervalState, time.Duration)) time.Duration {
+		for _, f := range checkFuncs {
+			state, newInterval := f()
+			if state == none {
```

Comment:
Should it be returned here if the `state` is `adaptiveUpdateTSIntervalStateUnadjustable` instead of continue processing?

Reply:
If state is not none, it should return at line 454